### ISSUE TITLE
No normalization

### ISSUE THREAD

**Issue by prubbens** [TarikExner/CytoNormPy#11]
I want to use CytoNormPy to normalize one sample in reference to another. 

I have formatted the metadata according to the example on this [page](https://cytonormpy.readthedocs.io/en/latest/vignettes/cnp_fcs_file.html). 

Everything runs successfully, however, when inspecting the normalized dataframes, everything remains the same, meaning the values are not normalized. `.calc_emd()` returns NaN-values. Plotting the histograms using the KDEs, returns no transformed distribution: 

```
cnpl = cnp.Plotter(cytonorm = cn)
cnpl.histogram(cn._datahandler.validation_file_names[0],
               x_channel = 'BL1-H',
               x_scale = "linear",
               y_scale = "linear",
               figsize = (4,4))
```

![afbeelding](https://github.com/user-attachments/assets/b980bb68-bc06-4094-8ed6-3d89bde3b54f)

Note that I did the preprocessing myself, so I didn't need a transformer. I also wrote a custom script to add csv-files into an anndata object, which is here below, and is based on this [page](https://cytonormpy.readthedocs.io/en/latest/vignettes/cnp_anndata.html): 

```
def _df_to_anndata(input_directory,
                   features,
                   file,
                   file_no,
                   metadata):
    df = pd.read_csv(input_directory + file, index_col=0, header=0)
    events = df.loc[:,features].values
    md_row = metadata.loc[metadata["file_name"] == file, :].to_numpy()
    obs = np.repeat(md_row, events.shape[0], axis = 0)
    var_frame =  pd.DataFrame({"pns": features,
                               "channel_numbers": np.arange(len(features))},
                              index = features)
    obs_frame = pd.DataFrame(
        data = obs,
        columns = metadata.columns,
        index = pd.Index([f"{file_no}-{str(i)}" for i in range(events.shape[0])])
    )
    adata = ad.AnnData(
        obs = obs_frame,
        var = var_frame,
        layers = {"compensated": events}
    )
    adata.obs_names_make_unique()
    adata.var_names_make_unique()
    return adata
```

Setting up the anndata object goes like this: 

```
adatas = []
input_directory = 'data/processed/'
for file_no, file in enumerate(filename):
    adatas.append(
        _df_to_anndata(input_directory, FEATURES, file, file_no, metadata)
    )
dataset = ad.concat(adatas, axis = 0, join = "outer", merge = "same")
dataset.obs = dataset.obs.astype("object")
dataset.var = dataset.var.astype("object")
dataset.obs_names_make_unique()
dataset.var_names_make_unique()
```

After which I run CytoNormPy as follows: 

```
cn = cnp.CytoNorm()
clusterer = cnp.FlowSOM(cols_to_use = FEATURES, n_clusters=25, rlen=100)
# clusterer = cnp.KMeans(n_clusters=10)
cn.add_clusterer(clusterer)
cn.run_anndata_setup(dataset, layer='compensated', channels=FEATURES, key_added='normalized')
cn.run_clustering(test_cluster_cv=True, cluster_cv_threshold=2)
cn.calculate_quantiles(n_quantiles=101)
cn.calculate_splines(goal = 'batch_median')
cn.normalize_data()
```

Any idea what the reason of this could be? 

Using the KMeans algorithm results in the same behavior.

**Comment by prubbens** [comment]
I subsampled to the .csv files I'd like to integrate to 1,000 cells each, along with the metadata. Please find them in attachment. And thanks for looking into this! 

     filename = ['df_brxx_stand_subsample.csv','df_bvxx_stand_subsample.csv']
     FEATURES = ['FSC-H','SSC-H','BL1-H','BL2-H','BL3-H','BL4-H']

[df_bvxx_stand_subsample.csv](https://github.com/user-attachments/files/17202069/df_bvxx_stand_subsample.csv)
[df_brxx_stand_subsample.csv](https://github.com/user-attachments/files/17202070/df_brxx_stand_subsample.csv)
[metadata_cytonorm_subsample.csv](https://github.com/user-attachments/files/17202073/metadata_cytonorm_subsample.csv)

**Comment by TarikExner** [comment]
Hey,

so that was a conceptual issue. There is no batch-normalization if you only have 1 batch. From what I understand what you want to do, you would have to set the files up as individual batches, so batch1 for brxx and batch2 for bvxx. Then proceed as usual.

I added the necessary code so that this approach doesn't fail and merged it into main already, so just re-download it and it should work if you have a single batch per file (set the reference value to "ref" for both).

If you encounter any different problems or if this was not what you had in mind let me know!

Best,
Tarik

**Comment by prubbens** [comment]
Thanks for looking into this! I downloaded the newest version. 

If I setup the metadata as suggested, I get the error `ValueError: Please make sure that there are only two values in the reference column. Have found ['ref']`

```
`filename = ['df_brxx_stand.csv','df_bvxx_stand.csv']
reference = ['ref','ref']
batch = [1,2]
sample_id = [1,2]
metadata = pd.DataFrame({'file_name':filename,
                         'reference':reference,
                         'batch':batch,
                         'sample_ID':sample_id})
```

If I set it up as follows, I get the following error: `ValueError: All batches must have reference samples.`

```
reference = ['other','ref']
batch = [1,2]
sample_id = [1,2]
metadata = pd.DataFrame({'file_name':filename,
                         'reference':reference,
                         'batch':batch,
                         'sample_ID':sample_id})
```

If I then setup the metadata as I originally did, I have no normalization. 

Basically I would like to normalize df_brxx based on df_bvxx (or the other way around). Am I misunderstanding something, or is there still a bug in the code?

**Comment by TarikExner** [comment]
Could you check manually in cytonormpy/_utils/_utils.py line 303 if it correctly states if len(_df[reference]) > 2?

I remember this from yesterday, but I adjusted this part.

Your first approach with two batches and 1 ref sample per batch is the correct one.

Best,
Tarik

**Comment by prubbens** [comment]
Okay, I had to uninstall and reinstall the package. 

I get a different dataframe back, so the normalization has taken place. One thing now is that `cn._datahandler.validation_file_names` seems empty, making it not possible to use: 

```
cnpl = cnp.Plotter(cytonorm = cn)

cnpl.histogram(cn._datahandler.validation_file_names[0],
               x_channel = 'BL1-H',
               x_scale = "linear",
               y_scale = "linear",
               figsize = (4,4))
```

**Comment by TarikExner** [comment]
Hey,

so this access of private attributes has to be changed in the vignettes at some point, I agree. For now, you can just use the filenames as they are!

The underlying plotting code uses the filename string to look up the respective file, so there is no reliance on the attribute validation_file_names.

Does this conclude this issue?

Best,
Tarik

**Comment by prubbens** [comment]
Check. Used to work with pandas dataframes, the `anndata` objects take a bit of time to getting used to. 

And I'd say so, thanks for your help!

**Comment by TarikExner** [comment]
No problem, youre welcome!


Best,
Tarik

**Comment by prubbens** [comment]
Question, still related to this issue. The resulting normalization after running CNP is a shared one, meaning that both `df_brxx` and `df_bvxx` are shifted. How can I setup the metadata so that one dataframe is corrected using the other? 

Thanks, 

Peter

**Comment by TarikExner** [comment]
Uhm, could you share the code that you used (I still have the csv files). 

That shouldn't happen actually, because since there is only one reference file, the spline functions are practically identity functions for the reference file, which means that the reference file will be mapped 1:1 and shouldnt shift...


Best,
Tarik

**Comment by prubbens** [comment]
I don't think it has changed compared to the code in this issue: 

```
filename = ['df_beads_vn_stand.csv','df_beads_be_stand.csv']
reference = ['ref','ref']
batch = [1,2]
sample_id = [1,2]
metadata = pd.DataFrame({'file_name':filename,
                         'reference':reference,
                         'batch':batch,
                         'sample_ID':sample_id})
metadata.to_csv('data/processed/metadata_beads_cytonorm.csv')
```

# Write custom function to transform pandas dataframe files to anndata

```
def _df_to_anndata(input_directory,
                   features,
                   file,
                   file_no,
                   metadata):
    df = pd.read_csv(input_directory + file, index_col=0, header=0)
    events = df.loc[:,features].values
    md_row = metadata.loc[metadata["file_name"] == file, :].to_numpy()
    obs = np.repeat(md_row, events.shape[0], axis = 0)
    var_frame =  pd.DataFrame({"pns": features,
                               "channel_numbers": np.arange(len(features))},
                              index = features)
    obs_frame = pd.DataFrame(
        data = obs,
        columns = metadata.columns,
        index = pd.Index([f"{file_no}-{str(i)}" for i in range(events.shape[0])])
    )
    adata = ad.AnnData(
        obs = obs_frame,
        var = var_frame,
        layers = {"compensated": events}
    )
    adata.obs_names_make_unique()
    adata.var_names_make_unique()
    return adata
```

```
# Create anndata files

adatas = []
input_directory = 'data/processed/'
for file_no, file in enumerate(filename):
    adatas.append(
        _df_to_anndata(input_directory, FEATURES, file, file_no, metadata)
    )
dataset = ad.concat(adatas, axis = 0, join = "outer", merge = "same")
dataset.obs = dataset.obs.astype("object")
dataset.var = dataset.var.astype("object")
dataset.obs_names_make_unique()
dataset.var_names_make_unique()
```

```
cn = cnp.CytoNorm()
clusterer = cnp.FlowSOM(cols_to_use = FEATURES, n_clusters=25, rlen=100)
# clusterer = cnp.KMeans(n_clusters=10)
cn.add_clusterer(clusterer)
cn.run_anndata_setup(dataset, layer='compensated', channels=FEATURES, key_added='normalized')
cn.run_clustering(test_cluster_cv=True, cluster_cv_threshold=2)
cn.calculate_quantiles(n_quantiles=101)
cn.calculate_splines(goal = 'batch_median')
cn.normalize_data()
cn.calculate_emd()
```

**Comment by TarikExner** [comment]
Hey Peter,

I am super sorry for the delay!

The easiest solution for this would be to use your setup and pass file_names and batch to cn.normalize_data I feel like.

So you run the pipeline as usual and then run
```python
cn.normalize_data(file_names = ['df_bvxx_stand_subsample.csv'], batches = [2])
file1 = adata[adata.obs["sample_ID"] == 1].copy()
file2 = adata[adata.obs["sample_ID"] == 2].copy()
assert np.array_equal(file1.layers["compensated"], file1.layers["normalized"])
assert not np.array_equal(file2.layers["compensated"], file2.layers["normalized"])
```
Note that I use the filenames that you initially uploaded, your updated code has other filenames. There is currently no error thrown if the file_name does not match any of the present files, I will add this later.

That should leave the first file unaltered while the second is transformed according to the spline function that you calculated between both files.

Let me know if this solves the problem!

Best, Tarik

**Comment by prubbens** [comment]
Yes, this seems to work, thanks. To keep the code consistent with the first example, I had to replace`adata`with `dataset`: 

`cn.normalize_data(file_names = ['df_bvxx_stand_subsample.csv'], batches = [2])
file1 = dataset[dataset.obs["sample_ID"] == 1].copy()
file2 = dataset[dataset.obs["sample_ID"] == 2].copy()
assert np.array_equal(file1.layers["compensated"], file1.layers["normalized"])
assert not np.array_equal(file2.layers["compensated"], file2.layers["normalized"])`

**Comment by TarikExner** [comment]
Happy to hear that it works!

Best,
Tarik
